src/lib.rs

```diff
@@ -116,40 +117,60 @@ pub struct Rules {
 }
 
 impl Rules {
-    /// Split a text into segments.
-    pub fn split<'a>(&self, text: &'a str) -> Vec<&'a str> {
+    /// Obtain the ranges for text segments. Guaranteed to be at character bounds.
+    pub fn split_ranges(&self, text: &str) -> Vec<Range<usize>> {
         let mut segments = Vec::new();
 
-        let mut mask: Vec<Option<bool>> = vec![None; text.len()];
+        // Mask defines the segments of text which are
+        // TODO use a proper tri-state enum here
+        let mut masked_bytes: Vec<Option<bool>> = vec![None; text.len()];
 
-        for rule in &self.rules {
-            for index in rule.match_indices(text) {
-                if index >= text.len() {
-                    continue;
+        'outer: for rule in &self.rules {
+            for byte_range in rule.match_indices(text) {
+                // byte range of non-content chars
+                let Range { start, end } = byte_range;
+
+                debug_assert!(end <= text.len());
+                if start >= text.len() {
+                    continue 'outer;
                 }
 
-                if mask[index].is_none() {
-                    mask[index] = Some(rule.do_break());
+                let byte_index = byte_range.start;
+                if masked_bytes[byte_index].is_none() {
+                    masked_bytes[byte_index] = Some(rule.do_break());
                 }
             }
         }
 
-        let mut prev_index = 0;
+        let mut prev_byte_pos = 0;
 
-        for (i, mask_val) in (0..text.len()).zip(mask) {
-            if let Some(true) = mask_val {
-                segments.push(&text[prev_index..i]);
-                prev_index = i;
+        // Iterate over characters, we don't want no half characters in the output ranges
+        for ((byte_pos, _c), tristate_mask) in text.char_indices().zip(masked_bytes) {
+            if let Some(true) = tristate_mask {
+                segments.push(prev_byte_pos..byte_pos);
+                prev_byte_pos = byte_pos;
             }
         }
 
-        if prev_index != text.len() {
-            segments.push(&text[prev_index..]);
+        // Deal with the trailing element, which is by definition
+        // not required to be suffixed by a gap char.
+        if text[prev_byte_pos..].chars().count() > 0 {
```

Comment:
Why `.chars().count()` here? I think the check `prev_index != text.len()` is sound. It probably makes a minuscule difference in performance, I'm just curious.

Reply:
If the last char is a multibyte char, the equality will not hold anymore. As such `chars().next().is_some()` would be the most efficient way.

Reply:
If the last char is a multibyte char which has not yet been added, `prev_index != text.len()` would be true so it would be correctly added.

But I like the `chars().next().is_some()` now, it's more readable!